// @aspect-ratio/core/lib/aspect-ratio.js
/**
 *
 */
export class AspectRatio {
    constructor(options) {
        var _a, _b, _c;
        this.options = options;
        /**
         * Iframe where the new width & height is calculated
         */
        this.iframe = window.document.createElement('iframe');
        /**
         * Style element child of the Iframe which hold the aspect
         * ratio css instructions
         */
        this.style = window.document.createElement('style');
        /**
         * Div element child of the Iframe which get affected by
         * the aspect ratio css instructions
         */
        this.sizer = window.document.createElement('div');
        const { container, mask, ratio } = options || {};
        const isNotValidOptions = !container || !mask;
        if (isNotValidOptions) {
            throw new Error('No valid container or mask element provided.');
        }
        this.iframe.classList.add('aspect-ratio-iframe');
        container.classList.add('aspect-ratio-container');
        mask.classList.add('aspect-ratio-mask');
        // attach iframe
        container.prepend(this.iframe);
        // add style
        (_a = this.iframe.contentDocument) === null || _a === void 0 ? void 0 : _a.body.appendChild(this.style);
        // add the sizer div
        (_b = this.iframe.contentDocument) === null || _b === void 0 ? void 0 : _b.body.appendChild(this.sizer);
        // add resize listerner
        (_c = this.iframe.contentWindow) === null || _c === void 0 ? void 0 : _c.addEventListener('resize', () => {
            this.resizeMask();
        });
        this.ratio = ratio;
    }
    /**
     * @param {number} minWidth
     */
    set minWidth(minWidth) {
        this.iframe.setAttribute('width', minWidth.toString());
    }
    /**
     * @param {number} minHeight
     */
    set minHeight(minHeight) {
        this.iframe.setAttribute('height', minHeight.toString());
    }
    /**
     * @param {Ratio} ratio
     */
    set ratio(ratio) {
        if (!ratio) {
            throw new Error('No valid ratio provided.');
        }
        if (typeof ratio === 'string') {
            this.updateRatioFromString(ratio);
        }
        if (typeof ratio === 'object') {
            this.updateRatioFromRatio(ratio);
        }
        this.updateStyle();
    }
    /**
     * @param {HTMLElement} mask
     */
    set mask(mask) {
        const isNotValidMask = !mask;
        if (isNotValidMask) {
            throw new Error('No valid mask element provided.');
        }
        this.options.mask = mask;
        this.resizeMask();
    }
    destroy() {
        this.iframe.remove();
    }
    updateStyle() {
        const { ratio, align } = this.options;
        const { numerator, denominator } = ratio;
        this.sizer.setAttribute('data-align', align);
        this.style.textContent = `
      html, body {
        margin: 0;
        padding: 0;
        background: transparent;
      }
      div {
        margin: auto;
        position: absolute;
        --numerator: ${numerator};
        --denominator: ${denominator};
        max-width: 100%;
        max-height: 100%;
        height: calc(1vw * var(--denominator) / var(--numerator) * 100);
        width: calc(1vh * var(--numerator) / var(--denominator) * 100);
      }

      div[data-align="center center"] {
        top: 0;
        left: 0;
        bottom: 0;
        right: 0;
      }
      div[data-align="center"] {
        left: 0;
        right: 0;
      }
      div[data-align="center top"] {
        left: 0;
        right: 0;
        top: 0;
      }
      div[data-align="center bottom"] {
        left: 0;
        right: 0;
        bottom: 0;
      }
    `;
        this.resizeMask();
    }
    resizeMask() {
        const coords = this.sizer.getBoundingClientRect();
        const { mask } = this.options;
        mask.style.width = coords.width + 'px';
        mask.style.height = coords.height + 'px';
        mask.style.top = coords.top + 'px';
        mask.style.bottom = coords.bottom + 'px';
        mask.style.left = coords.left + 'px';
        mask.style.right = coords.right + 'px';
    }
    updateRatioFromString(ratioStr) {
        const [numeratorStr, denominatorStr] = (ratioStr || '16/9').split('/');
        const numerator = parseInt(numeratorStr, 10);
        const denominator = parseInt(denominatorStr, 10);
        const ratio = { denominator, numerator };
        if (!this.isValidRatio(ratio))
            return;
        this.options = Object.assign(Object.assign({}, this.options), { ratio });
    }
    updateRatioFromRatio(ratio) {
        if (!this.isValidRatio(ratio))
            return;
        this.options = Object.assign(Object.assign({}, this.options), { ratio });
    }
    isValidRatio({ numerator, denominator }) {
        const isBiggerThan0 = (n) => n > 0;
        return isBiggerThan0(numerator) && isBiggerThan0(denominator);
    }
}
